toolchain/parse/extract.cpp

```diff
@@ -0,0 +1,341 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <tuple>
+#include <utility>
+
+#include "common/error.h"
+#include "common/struct_reflection.h"
+#include "toolchain/parse/tree.h"
+#include "toolchain/parse/typed_nodes.h"
+
+namespace Carbon::Parse {
+
+// A trait type that should be specialized by types that can be extracted
+// from a parse tree. A specialization should provide the following API:
+//
+// ```
+// interface Extractable {
+//   // Extract a value of this type from the sequence of nodes starting at
+//   // `it`, and increment `it` past this type. Returns `std::nullopt` if
+//   // the tree is malformed. If `trace != nullptr`, writes what actions
+//   // were taken to `*trace`.
+//   static auto Extract(Tree* tree, Tree::SiblingIterator& it,
+//                       Tree::SiblingIterator end,
+//                       ErrorBuilder* trace) -> std::optional<Self>;
+// }
+// ```
```

Comment:
While I love the Carbon code in here, I think maybe it's better to use a more direct C++ example.

```suggestion
// ```cpp
// template<>
// struct Extractable<T> {
//   // Extract a value of this type from the sequence of nodes starting at
//   // `it`, and increment `it` past this type. Returns `std::nullopt` if
//   // the tree is malformed. If `trace != nullptr`, writes what actions
//   // were taken to `*trace`.
//   static auto Extract(Tree* tree, Tree::SiblingIterator& it,
//                       Tree::SiblingIterator end,
//                       ErrorBuilder* trace) -> std::optional<T>;
// };
// ```
```

Reply:
Honestly, not sure its an improvement... the Cabron feature here is really nice for illustrating what is going on...

So consider this optional.